File: neet_code_all/src/stack/remove_all_adjacent_duplicates_in_string_2.rs
// 解けなかった
struct Solution;
impl Solution {
    pub fn remove_duplicates(s: String, k: i32) -> String {
        let mut stack = vec![];
        for c in s.chars() {
            stack.push(c);
            let mut len_stack = stack.len();
            for i in (0..len_stack).rev() {}
        }

        stack.iter().collect::<String>()
    }
}

// 模範解答
struct SolutionAns;
impl SolutionAns {
    pub fn remove_duplicates(s: String, k: i32) -> String {
        let mut stack: Vec<(char, usize)> = vec![];
        for c in s.chars() {
            if !stack.is_empty() && stack.last().unwrap().0 == c {
                let mut last = stack.pop().unwrap();
                last.1 += 1;
                stack.push(last);
            } else {
                stack.push((c, 1));
            }
            if stack.last().unwrap().1 == k as usize {
                stack.pop();
            }
        }

        stack.iter().fold(String::new(), |acc, &(c, count)| {
            acc + &c.to_string().repeat(count)
        })
    }

    // AC
    pub fn remove_duplicates_2(s: String, k: i32) -> String {
        let mut stack: Vec<(char, usize)> = vec![];
        for c in s.chars() {
            if let Some(stack_last) = stack.last() {
                if stack_last.0 == c {
                    let mut last = stack.pop().unwrap();
                    last.1 += 1;
                    stack.push(last);
                } else {
                    stack.push((c, 1));
                }
            } else {
                stack.push((c, 1));
            }

            if let Some(stack_last) = stack.last() {
                if stack_last.1 == k as usize {
                    stack.pop();
                }
            }
        }

        stack.iter().fold(String::new(), |acc, &(c, count)| {
            acc + &c.to_string().repeat(count)
        })
    }
}

fn main() {
    let case_1 = ("abcd".to_string(), 2);
    // => "abcd"
    let case_2 = ("deeedbbcccbdaa".to_string(), 3);
    // => "aa"
    let case_3 = ("pbbcggttciiippooaais".to_string(), 2);
    // => "ps"

    /*
    println!(
        "case_2: {}",
        Solution::remove_duplicates(case_2.0.clone(), case_2.1)
    )
    */

    println!(
        "case_1: {}",
        SolutionAns::remove_duplicates(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAns::remove_duplicates(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAns::remove_duplicates(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_1: {}",
        SolutionAns::remove_duplicates_2(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAns::remove_duplicates_2(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAns::remove_duplicates_2(case_3.0.clone(), case_3.1)
    );
}

File: neet_code_all_cpp/stack/remove_all_adjacent_duplicates_in_string_2.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <string>
#include <vector>

// 模範解答
class SolutionAns {
    public:
        std::string removeDuplicates(std::string s, int k) {
            std::stack<std::pair<char, int>> stack;

            for(auto& c: s) {
                int count = 1;
                if(!stack.empty() && stack.top().first == c) {
                    count += stack.top().second;
                    stack.pop();
                }

                stack.push({c, count});
                if(count == k) {
                    stack.pop();
                }
            }

            std::string result = "";
            while(!stack.empty()) {
                auto [c, freq] = stack.top();
                while(freq > 0) {
                    result += c;
                    --freq;
                }

                stack.pop();
            } 

            std::reverse(result.begin(), result.end());
            return result;
        }
};

// Rustの模範解答より
// AC
class SolutionAnsRust {
    public:
        std::string removeDuplicates(std::string s, int k) {
            std::vector<std::pair<char, int>> stack;
            for(auto&c : s) {
                if(!stack.empty() && stack.back().first == c) {
                    auto last = stack.back();
                    stack.pop_back();
                    ++last.second;
                    stack.push_back(last);
                } else {
                    stack.push_back({c, 1});
                }
                if(stack.back().second == k) {
                    stack.pop_back();
                }
            }

            std::string result = "";
            for(auto& [c, freq]: stack) {
                for(int i = 0; i < freq; ++i) {
                    result.push_back(c);
                }
            }
            return result;
        }
};

int main(void) {
    using Case = std::pair<std::string, int>;
    Case case_1 = {"abcd", 2};
    // => "abcd"
    Case case_2 = {"deeedbbcccbdaa", 3};
    // => "aa"
    Case case_3 = {"pbbcggttciiippooaais", 2};
    // => "ps"

    SolutionAns s_ans;
    std::cout << s_ans.removeDuplicates(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.removeDuplicates(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.removeDuplicates(case_3.first, case_3.second) << std::endl;

    SolutionAnsRust s_ans_rs;
    std::cout << s_ans_rs.removeDuplicates(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans_rs.removeDuplicates(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans_rs.removeDuplicates(case_3.first, case_3.second) << std::endl;
}
